modules/caddyhttp/metrics.go

```diff
@@ -0,0 +1,215 @@
+package caddyhttp
+
+import (
+	"context"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var httpMetrics = struct {
+	init                   sync.Once
+	requestInFlight        *prometheus.GaugeVec
+	requestCount           *prometheus.CounterVec
+	requestErrors          *prometheus.CounterVec
+	requestDuration        *prometheus.HistogramVec
+	requestDurationSummary *prometheus.SummaryVec
+	requestSize            *prometheus.HistogramVec
+	responseSize           *prometheus.HistogramVec
+	responseLatency        *prometheus.HistogramVec
+	responseLatencySummary *prometheus.SummaryVec
+}{
+	init: sync.Once{},
+}
+
+func initHTTPMetrics() {
+	ns := "caddy"
+	sub := "http"
+
+	basicLabels := []string{"server", "handler"}
+	httpMetrics.requestInFlight = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: ns,
+		Subsystem: sub,
+		Name:      "requests_in_flight",
+		Help:      "Number of requests currently handled by this server.",
+	}, basicLabels)
+	httpMetrics.requestErrors = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: ns,
+		Subsystem: sub,
+		Name:      "request_errors_total",
+		Help:      "Number of requests resulting in middleware errors.",
+	}, basicLabels)
+	httpMetrics.requestCount = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: ns,
+		Subsystem: sub,
+		Name:      "requests_total",
+		Help:      "Counter of HTTP(S) requests made.",
+	}, basicLabels)
+
+	// TODO: allow these to be customized in the config
+	latencyBuckets := []float64{.01, .05, .1, .2, .4, 1, 3, 8, 20, 60, 120}
```

Comment:
I'd recommend starting with `prometheus.DefBuckets` as the default.

Reply:
For ref: https://pkg.go.dev/github.com/prometheus/client_golang/prometheus?tab=doc#pkg-variables

Reply:
@SuperQ sure. I started with that, but wasn't happy with the spread. But I think this is going to need to be configurable anyway.